pox-locking/src/pox_4.rs

```diff
@@ -0,0 +1,395 @@
+    global_context: &mut GlobalContext,
+    function_name: &str,
+    value: &Value,
+) -> Result<Option<StacksTransactionEvent>, ClarityError> {
+    // in this branch case, the PoX-4 contract has stored the increase information
+    //  and performed the increase checks. Now, the VM needs to update the account locks
+    //  (because the locks cannot be applied directly from the Clarity code itself)
+    // applying a pox lock at this point is equivalent to evaluating a transfer
+    debug!(
+        "Handle special-case contract-call";
+        "contract" => ?boot_code_id("pox-4", global_context.mainnet),
+        "function" => function_name,
+        "return-value" => %value,
+    );
+
+    runtime_cost(
+        ClarityCostFunction::StxTransfer,
+        &mut global_context.cost_track,
+        1,
+    )?;
+
+    let (stacker, total_locked) = match parse_pox_increase(value) {
+        Ok(x) => x,
+        Err(_) => {
+            // nothing to do -- function failed
+            return Ok(None);
+        }
+    };
+    match pox_lock_increase_v4(&mut global_context.database, &stacker, total_locked) {
+        Ok(new_balance) => {
+            let event =
+                StacksTransactionEvent::STXEvent(STXEventType::STXLockEvent(STXLockEventData {
+                    locked_amount: new_balance.amount_locked(),
+                    unlock_height: new_balance.unlock_height(),
+                    locked_address: stacker,
+                    contract_identifier: boot_code_id(POX_4_NAME, global_context.mainnet),
+                }));
+
+            Ok(Some(event))
+        }
+        Err(LockingError::DefunctPoxContract) => Err(ClarityError::Runtime(
+            RuntimeErrorType::DefunctPoxContract,
+            None,
+        )),
+        Err(e) => {
+            // Error results *other* than a DefunctPoxContract panic, because
+            //  those errors should have been caught by the PoX contract before
+            //  getting to this code path.
+            panic!("FATAL: failed to increase lock from {stacker}: '{e:?}'");
+        }
+    }
+}
+
+/// Handle special cases when calling into the PoX-4 API contract
+pub fn handle_contract_call(
+    global_context: &mut GlobalContext,
+    sender_opt: Option<&PrincipalData>,
+    contract_id: &QualifiedContractIdentifier,
+    function_name: &str,
+    args: &[Value],
+    value: &Value,
+) -> Result<(), ClarityError> {
+    // Generate a synthetic print event for all functions that alter stacking state
+    let print_event_opt = if let Value::Response(response) = value {
+        if response.committed {
+            // method succeeded.  Synthesize event info, but default to no event report if we fail
+            // for some reason.
+            // Failure to synthesize an event due to a bug is *NOT* an excuse to crash the whole
+            // network!  Event capture is not consensus-critical.
+            let event_info_opt = match synthesize_pox_2_or_3_event_info(
```

Comment:
This function `synthesize_pox_2_or_3_event_info()` should be named something more informative, now that pox-4 exists.  Perhaps `synthesize_pox_event_info()`?

Reply:
I like the name. If we agree on it, I'll update the PR #4094 that changes the pox-lockup::events

Reply:
Done in https://github.com/stacks-network/stacks-core/pull/4094/commits/915fba782da0009cdef509a0522bdc0476c11265